model/src/main/kotlin/Provenance.kt

```diff
@@ -97,12 +97,11 @@ data class Provenance(
         }
 
         return listOf(pkg.vcs, pkg.vcsProcessed).any {
-            // If "it" has a resolved revision it must be equal to the resolved revision of vcsInfo, otherwise the
-            // revision of "it" has to equal either the revision or the resolved revision of vcsInfo.
-            it.type == vcsInfo.type && it.url == vcsInfo.url && it.path == vcsInfo.path
-                    && (it.resolvedRevision?.let {
-                vcsInfo.resolvedRevision == it
-            } ?: vcsInfo.resolvedRevision == it.revision || vcsInfo.revision == it.revision)
+            if (it.resolvedRevision != null) {
+                it.resolvedRevision == vcsInfo.resolvedRevision
+            } else {
+                vcsInfo.resolvedRevision == it.revision || vcsInfo.revision == it.revision
```

Comment:
Shouldn't the `|| vcsInfo.revision == it.revision` part be moved out of the `if` / `else`? I was thinking to rewrite the code as

            it.type == vcsInfo.type && it.url == vcsInfo.url && it.path == vcsInfo.path && (
                    vcsInfo.resolvedRevision == it.resolvedRevision ?: it.revision ||
                    vcsInfo.revision == it.revision
            )

Reply:
Hm, why should it be moved out?
I think it now aligns as is with your code comment I removed?!

Reply:
Looking at your code again it indeed seems to be correct, although I still find it confusing. For my thinking I had rewritten only the code to compare the (resolved) revisions as follows:

    (it.resolvedRevision != null && vcsInfo.resolvedRevision == it.resolvedRevision) ||
    (it.resolvedRevision == null && vcsInfo.resolvedRevision == it.revision) ||
    vcsInfo.revision == it.revision

I.e. all these three *or*-ed conditions are equal. But in your code it looks like the latter two conditions need to go hand-in-hand somehow. Maybe it's clearer to replace

    vcsInfo.resolvedRevision == it.revision || vcsInfo.revision == it.revision

with

    it.revision in listOf(vcsInfo.resolvedRevision, vcsInfo.revision)

in your new code?

Reply:
Also, re-reading the comment again, I noticed that the order of "revision" vs. "resolved revision" differs in docs vs code. Although it probably does not matter in practice, we should make that match, too, by either adjusting the code or the docs.

Reply:
to me all three approaches are equivalent in terms of the return value, a I missing something?:
- it.revision == vcsInfo.revision || it.revision == vcsInfo.resolvedRevision
- it.revision in listOf(vcsInfo.revision, vcsInfo.resolvedRevision)
- vcsInfo.resolvedRevision == it.revision || vcsInfo.revision == it.revision  (my code)

Reply:
Yes, they should be equal in this case. However, with the exact wording of the original code comment in mind, I find `it.revision == vcsInfo.revision || it.revision == vcsInfo.resolvedRevision` to be the clearest / best match (I also like the list approach for readability, but creating a temporary list here is probably overkill).

The reason why I'm picky about the order is also because it expresses some sort o expectation / can be used to optimize code (though not in this case): If the first *or*-condition matches, the second one is not even evaluated. If the individual conditions involved (expensive) functions calls, it makes sense to check the least expensive but most likely condition first. But I agree this is a tad academic and is not really relevant in this case.

Reply:
> The reason why I'm picky about the order is also because it expresses some sort o expectation / can be used to optimize code (though not in this case): If the first _or_-condition matches, the second one is not even evaluated. If the individual conditions involved (expensive) functions calls, it makes sense to check the least expensive but most likely condition first. But I agree this is a tad academic and is not really relevant in this case.

I think I missed this information for understanding why you were asking to change it.
I adjusted it and I find it very slightly clearer ;-)